# Lab 5: Dictionaries 

In this lab you will practice the use of dictionaries. Do the following tasks. 

**Problem 1: Reading a file of data**
In the module there is a file named "Birth_Life.txt". You will need to download this file and put it into the same working directory you have decided to save this lab file to for easy access to the file. You should open this file and review its contents but make no changes to the file. You should see it is information about a selection of countries. Specifically, Country Name is a column that contains the country's name, followed by Life_Exp this is the average life time in years for that country, followed by Birth_Rate this is the average birthrate for that country per 1000 people in a year. 

1. Using the open("filename", "r") function read the file and store the output into a variable named f_obj (file object). 
2. now use the file_object.read() function to read all of the information, and store it in alldata variable. The .read() function will essentially read all of the data into a single line of information. 
3. Split the data using alldata.split("\n") this will split the data list based on the new line marker locations, which is at the end of each line of information in the original text. You should store the output of this into a new variable called alldata_split. 
4. This is where my help ends and the task really begins. Explore the current output that you have stored into alldata_split variable from point 3. Using a for loop, range, and split, create a list of lists, where all the data from one line is put into a list. the desired output should look like this for the first two lines: 
[['Country', 'Name', 'Life_Exp', 'Birth_Rate', 'GNI', 'Development'],
 ['"Congo,', 'Dem.', 'Rep."', '50.00', '45.96', '130', 'Developing'],...]
 
This list of lists should be named data_list. Print the output of the code meaning, print the data_list variable on to the console. 

5. The last thing close the file object, run this line of code at the end `f_obj.close()`.

In [29]:
f_obj = open("Birth_Life.txt", "r")

alldata = f_obj.read()
alldata_split = alldata.split("\n")

data_list = []

for line in alldata_split:
    split_line = line.split("\t")
    
    data_list.append(split_line)

print(data_list)

f_obj.close()

[['Country_Name', 'Life_Exp', 'Birth_Rate', 'GNI', 'Development'], ['Congo_Dem_Rep', '50.00', '45.96', '130', 'Developing'], ['India', '62.59', '26.46', '440', 'Developing'], ['South_Africa', '56.37', '24.33', '3070', 'Developing'], ['China', '72.00', '14.03', '940', 'Developing'], ['Pakistan', '62.76', '32.04', '490', 'Developing'], ['Mexico', '74.30', '24.13', '5810', 'Developing'], ['United_States', '76.64', '14.40', '36070', 'Developed'], ['Germany', '77.93', '9.30', '26210', 'Developed'], ['Switzerland', '79.68', '10.90', '43460', 'Developed'], ['Japan', '81.08', '9.40', '36230', 'Developed']]


**Problem 2: Making a Dictionary**
In this part of the assignment the goal is now to take the list of lists or the data_list variable contents and begin to construct a dictionary. For example, first row of the data_list is the keys for each value. For example for the Japan or last line of the file. The dictionary should be {'Country_Name':'Japan','Life_Exp':81.08 and so on. This will require that numerical data be converted from text into actual data. The goal here is use all that you have learned to create the dictionary. Hint, the easiest way to to this is two create a dictionary from two list. 
The desired output should look like: 

```
{'Country_1': {'Country_Name': 'Congo_Dem_Rep',
  'Life_Exp': 50.0,
  'Birth_Rate': 45.96,
  'GNI': 130.0,
  'Development': 'Developing'},
 'Country_2': {'Country_Name': 'India',
  'Life_Exp': 62.59,
  'Birth_Rate': 26.46,
  'GNI': 440.0,
  'Development': 'Developing'},...
 ```
 You must generate a name, thus you are creating a dictionary of dictionaries. This should use one for loop. Lastly, print the dictionary as the final task for this problem. 
 
*Hint: This may require that you stop and reload the file multiple times as you work through the code, you may change the structure of the data_list and you may need to refresh it by reading in the file again.* 

In [98]:
filename = 'Birth_Life.txt'

file = open(filename, 'r')

data_list = [line.strip().split('\t') for line in file.readlines()]

file.close()

headers = data_list[0]
country_data = {
    f'Country_{i}': {
        'Country_Name': row[0],
        'Life_Exp': float(row[1]),
        'Birth_Rate': float(row[2]),
        'GNI': float(row[3]),
        'Development': row[4]
    }
    for i, row in enumerate(data_list[1:], start=1)
}

print("{")
for country, data in country_data.items():
    print(f"  {country}: {{")
    for key, value in data.items():
        print(f"'{key}': {value},")
    print("  },")

{
  Country_1: {
'Country_Name': Congo_Dem_Rep,
'Life_Exp': 50.0,
'Birth_Rate': 45.96,
'GNI': 130.0,
'Development': Developing,
  },
  Country_2: {
'Country_Name': India,
'Life_Exp': 62.59,
'Birth_Rate': 26.46,
'GNI': 440.0,
'Development': Developing,
  },
  Country_3: {
'Country_Name': South_Africa,
'Life_Exp': 56.37,
'Birth_Rate': 24.33,
'GNI': 3070.0,
'Development': Developing,
  },
  Country_4: {
'Country_Name': China,
'Life_Exp': 72.0,
'Birth_Rate': 14.03,
'GNI': 940.0,
'Development': Developing,
  },
  Country_5: {
'Country_Name': Pakistan,
'Life_Exp': 62.76,
'Birth_Rate': 32.04,
'GNI': 490.0,
'Development': Developing,
  },
  Country_6: {
'Country_Name': Mexico,
'Life_Exp': 74.3,
'Birth_Rate': 24.13,
'GNI': 5810.0,
'Development': Developing,
  },
  Country_7: {
'Country_Name': United_States,
'Life_Exp': 76.64,
'Birth_Rate': 14.4,
'GNI': 36070.0,
'Development': Developed,
  },
  Country_8: {
'Country_Name': Germany,
'Life_Exp': 77.93,
'Birth_Rate': 9.3,
'GNI': 26210.0,
'Developme

**Problem 3: Knowledge is Power**
Hopefully by now you are seeing that your skills are growing. As such, now its time to pierce into the data to see what can be found. Write a interactive code that is capable of going through the dictionary you have built in the previous problem to provide a user with the result of the following options. Thus you must create the interactive code to present and obtain the choice of the user. Then you must use conditionals, prints, and computations to provide the result for each one. 

*HINT & ADVISE: It is advisable that you work on code for each part separate, test and assure it works. Then work on the combination all the code. I advise that you do not start with the input part first as then as you are trying to develop the code it will prompt you for an input. Do the code for each of the options first, then the main input. *

1. Would you like to know the data for a particular country? 
2. Would you like to know which contries are considered developing?
3. Would you like the average Birth_Rate, Life_Expectancy, and GNI for all developing countries? 
4. Would you like the average Birth_Rate, Life_Expectancy, and GNI for all developed countries? 
5. Would you like to know which contries are considered developed? 
6. Would you like the average Birth_Rate, Life_Expectancy, and GNI for all 10 countries?

For 1. You will prompt the user for which county. Thus, you should print a numbered list of the country names to choose from. You can take the country name or a number choice for an input, which is up to you to decide. Then use this input to print out the key and value pairs for the country chosen. Desired output:
```
Country_Name ---> Pakistan
Life_Exp ---> 62.76
Birth_Rate ---> 32.04
GNI ---> 490.0
Development ---> Developing
```
For 2. You should simply print out the countries that have "Developing" under the "Development" key in each country. The desired input should look like: 
```
The following countries are developing: 
Congo_Dem_Rep
India
South_Africa
China
Pakistan
Mexico
```
For 3. and 4. could be done in a single code segment/module where a choice of 3 or 4 is evaluated and will compute the average for developing and not developing at the same time, but you should assure the code fully tested for both cases. Desired output should look like:
```
The averages of developing countries are:
Average GNI ---> 1813.33
Average Birth Rate ---> 27.82
Average Life Expectancy ---> 63.0
```
For 5. should be the opposite of 2. and the output should look the same but for contries considered developed. 

For 6. the output should be similar to the one in 3 or 4 but should have values that correspond to the averages computed across all 10 countries. 



In [6]:
filename = 'Birth_Life.txt'

file = open(filename, 'r')
data_list = [line.strip().split('\t') for line in file.readlines()]
file.close()

headers = data_list[0]
country_data = {row[0]: {'Life_Exp': float(row[1]), 'Birth_Rate': float(row[2]), 'GNI': float(row[3]), 'Development': row[4]} for row in data_list[1:]}

while True:
    print("\n1. Get country data\n2. List developing countries\n3. Average for developing countries\n4. Average for developed countries")
    print("5. List developed countries\n6. Average for all countries\n7. Exit")

    choice = input("Enter your choice: ")

    if choice == '1':
        countries = list(country_data.keys())
        for i, country in enumerate(countries, 1):
            print(f"{i}. {country}")

        selected_index = input("Select a country by number: ")
        if selected_index.isdigit() and 1 <= int(selected_index) <= len(countries):
            selected = countries[int(selected_index) - 1]
            data = country_data[selected]
            print(f"\nCountry: {selected}\nLife Exp: {data['Life_Exp']}\nBirth Rate: {data['Birth_Rate']}\nGNI: {data['GNI']}\nDevelopment: {data['Development']}")
        else:
            print("Invalid selection.")

    elif choice in {'2', '5'}:
        dev_type = 'Developing' if choice == '2' else 'Developed'
        print(f"\nThe following countries are {dev_type.lower()}:")
        print("\n".join(k for k, v in country_data.items() if v['Development'] == dev_type) or "None found.")

    elif choice in {'3', '4'}:
        dev_type = 'Developing' if choice == '3' else 'Developed'
        subset = [v for v in country_data.values() if v['Development'] == dev_type]
        if subset:
            print(f"\nAverages for {dev_type.lower()} countries:")
            print(f"GNI: {sum(c['GNI'] for c in subset) / len(subset):.2f}")
            print(f"Birth Rate: {sum(c['Birth_Rate'] for c in subset) / len(subset):.2f}")
            print(f"Life Exp: {sum(c['Life_Exp'] for c in subset) / len(subset):.2f}")
        else:
            print(f"No {dev_type.lower()} countries found.")

    elif choice == '6':
        print("\nAverages for all countries:")
        for key in ['GNI', 'Birth_Rate', 'Life_Exp']:
            print(f"{key.replace('_', ' ')}: {sum(v[key] for v in country_data.values()) / len(country_data):.2f}")

    elif choice == '7':
        print("Exiting.")
        break

    else:
        print("Invalid choice.")



1. Get country data
2. List developing countries
3. Average for developing countries
4. Average for developed countries
5. List developed countries
6. Average for all countries
7. Exit


Enter your choice:  asdf


Invalid choice.

1. Get country data
2. List developing countries
3. Average for developing countries
4. Average for developed countries
5. List developed countries
6. Average for all countries
7. Exit


Enter your choice:  


Invalid choice.

1. Get country data
2. List developing countries
3. Average for developing countries
4. Average for developed countries
5. List developed countries
6. Average for all countries
7. Exit


Enter your choice:  1


1. Congo_Dem_Rep
2. India
3. South_Africa
4. China
5. Pakistan
6. Mexico
7. United_States
8. Germany
9. Switzerland
10. Japan


Select a country by number:  7



Country: United_States
Life Exp: 76.64
Birth Rate: 14.4
GNI: 36070.0
Development: Developed

1. Get country data
2. List developing countries
3. Average for developing countries
4. Average for developed countries
5. List developed countries
6. Average for all countries
7. Exit


Enter your choice:  7


Exiting.


**Bonus Extra Credit (10pts towards exam):** Take your code, all of it, and run pylint on it. Take a screen shot and show the overall grade you got on your code (x/10). Now submit the code to chat GPT and ask it to format your code into pep8 standard format. Then run the ChatGPT output through pylint to see what grade ChatGPT would get. Is it better? Show the details or better yet to get the point stop by and show me the result. 